lib/auth/autoupdate/autoupdatev1/service.go

```diff
@@ -617,3 +635,126 @@ func checkAdminCloudAccess(authCtx *authz.Context) error {
 	}
 	return nil
 }
+
+// Those values are arbitrary, we will want to increase them as we test. We will also want to modulate them based on the
+// cluster context. We don't want people to craft schedules that can't realistically finish within a week on Cloud as
+// we usually do weekly updates. However, self-hosted users can craft more complex schedules, slower rollouts, and shoot
+// themselves in the foot if they want.
+const (
+	maxGroupsTimeBasedStrategy        = 20
+	maxGroupsHaltOnErrorStrategy      = 10
+	maxGroupsHaltOnErrorStrategyCloud = 4
+	maxRolloutDurationCloudHours      = 72
+)
+
+var (
+	cloudGroupUpdateDays = []string{"Mon", "Tue", "Wed", "Thu"}
+)
+
+// validateServerSideAgentConfig validates that the autoupdate_config.agent spec meets the cluster rules.
+// Rules may vary based on the cluster, and over time.
+//
+// This function should not be confused with api/types/autoupdate.ValidateAutoUpdateConfig which validates the integrity
+// of the resource and does not enforce potentially changing rules.
+func validateServerSideAgentConfig(config *autoupdate.AutoUpdateConfig) error {
+	// We must check resource integrity before, because it makes no sense to try to enforce rules on an invalid resource.
+	// The generic backend service will likely check integrity again, but it's not a large performance problem.
+	err := update.ValidateAutoUpdateConfig(config)
+	if err != nil {
+		return trace.Wrap(err, "validating autoupdate config")
+	}
+
+	var maxGroups int
+	isCloud := modules.GetModules().Features().Cloud
+
+	agentsSpec := config.GetSpec().GetAgents()
+	switch {
+	case isCloud && agentsSpec.GetStrategy() == update.AgentsStrategyHaltOnError:
+		maxGroups = maxGroupsHaltOnErrorStrategyCloud
+	case agentsSpec.GetStrategy() == update.AgentsStrategyHaltOnError:
+		maxGroups = maxGroupsHaltOnErrorStrategy
+	case agentsSpec.GetStrategy() == update.AgentsStrategyTimeBased:
+		maxGroups = maxGroupsTimeBasedStrategy
+	default:
+		return trace.BadParameter("unknown max group for strategy %v", agentsSpec.GetStrategy())
+	}
+
+	if len(agentsSpec.GetSchedules().GetRegular()) > maxGroups {
+		return trace.BadParameter("max groups (%d) exceeded for strategy %s, %s schedule contains %d groups", maxGroups, agentsSpec.GetStrategy(), update.AgentsScheduleRegular, len(agentsSpec.GetSchedules().GetRegular()))
+	}
+
+	if isCloud {
```

Comment:
```suggestion
	if !isCloud {
		return nil
	}
```
nit: cleans up indent a bit